crates/utils/src/utils/markdown/spoiler_rule.rs

```diff
@@ -0,0 +1,109 @@
+// Custom Markdown plugin to manage spoilers.
+//
+// Matches the capability described in Lemmy UI:
+// https://github.com/LemmyNet/lemmy-ui/blob/main/src/shared/utils.ts#L159
+// that is based off of:
+// https://github.com/markdown-it/markdown-it-container/tree/master#example
+//
+// FORMAT:
+// Input Markdown: ::: spoiler THE_LEAD_IN\nTHE_HIDDEN_SPOILER\n:::\n
+// Output HTML: <details><summary>THE_LEAD_IN</summary><p>THE_HIDDEN_SPOILER</p></details>
+
+use markdown_it::{
+  parser::{
+    block::{BlockRule, BlockState},
+    inline::InlineRoot,
+  },
+  MarkdownIt,
+  Node,
+  NodeValue,
+  Renderer,
+};
+use once_cell::sync::Lazy;
+use regex::Regex;
+
+#[derive(Debug)]
+struct SpoilerBlock {
+  call_to_action: String,
+}
+
+const SPOILER_PREFIX: &str = "::: spoiler ";
+const SPOILER_SUFFIX: &str = ":::";
+const SPOILER_SUFFIX_NEWLINE: &str = ":::\n";
+
+static SPOILER_REGEX: Lazy<Regex> =
+  Lazy::new(|| Regex::new(r"^::: spoiler .*$").expect("compile spoiler markdown regex."));
+
+impl NodeValue for SpoilerBlock {
+  fn render(&self, node: &Node, fmt: &mut dyn Renderer) {
+    fmt.cr();
+    fmt.open("details", &node.attrs);
+    fmt.open("summary", &[]);
+    fmt.text(&self.call_to_action); // Not allowing special styling here to keep it simple.
+    fmt.close("summary");
+    fmt.open("p", &[]);
+    fmt.contents(&node.children);
+    fmt.close("p");
+    fmt.close("details");
+    fmt.cr();
+  }
+}
+
+struct SpoilerBlockScanner;
+
+impl BlockRule for SpoilerBlockScanner {
+  fn run(state: &mut BlockState) -> Option<(Node, usize)> {
+    let first_line: &str = state.get_line(state.line).trim();
+
+    // Check if the first line contains the spoiler syntax...
+    if !SPOILER_REGEX.is_match(first_line) {
+      return None;
+    }
+
+    let begin_fence_line_idx: usize = state.line + 1;
+    let mut end_fence_line_idx: usize = begin_fence_line_idx;
+    let mut has_end_marker: bool = false;
+
+    // Search for the end of the spoiler; there could potentially be multiple lines between
+    // the beginning and end of the block.
+    //
+    // Block ends with a line with ':::' or ':::\n'; it must be isolated from other markdown.
+    while end_fence_line_idx < state.line_max && !has_end_marker {
```

Comment:
Should nested spoilers be supported? Will tweak this functionality and add a test if it should.